### PR TITLE
Feature/add quickopen

### PR DESCRIPTION
**Pull request by Akin909**
quickopen

**Update**: with input
quickopen-2

### Outstanding (for a separate PR: see revery-ui/revery#408):
There's still a few kinks to iron out which primarily relate to things in revery
so I'd like to tie those up and then return to this in a separate PR once those changes are in.

- [x] Input should autofocus on open
- [x] Expose the key pressed (in Revery) so that logic based on the specific key can be handled in the onchange - use this to handle pressing escape which should return focus to Oni2
- [x] Fix cursor positioning on start
- [x] Add arrow key behaviour to input
- [ ] Fix backspace not responding

### CODE REVIEW

**Comment by Akin909** [comment]
@bryphe was getting pretty carried away trying to do this all in one PR but there needs to be a fair bit of work on the input component in revery to fix some of the outstanding issues I laid out in the description.

I think what this PR adds is a solid start and the next few things are polishes

File: assets/configuration/keybindings.json [hunk 1]
```diff
@@ -1,9 +1,10 @@
 {
     "bindings": [
+        { "key": "<S-C-P>", "command": "quickOpen.open", "when": ["editorTextFocus"] },
```
Comment (bryphe) [inline assets/configuration/keybindings.json]: These bindings - the `<S-C-P>` for quick open, and `<C-P>` for command palette seem reversed to me (compared both to Oni1 or VSCode)

File: src/editor/Model/Menu.re [hunk 2]
```diff
@@ -0,0 +1,62 @@
+open Oni_Core;
+open Types;
+open UiMenu;
+
+let create = (~effects: option(Effects.t)=?, ()) => {
+  menu: Closed,
```
Comment (bryphe) [inline src/editor/Model/Menu.re]: It's not clear to me what `menu` is for - where does it get set to something besides `Closed`?

File: src/editor/Model/QuickOpen.re [hunk 3]
```diff
@@ -0,0 +1,24 @@
+open Oni_Core.Types;
+open UiMenu;
+
+let content = (effects: Effects.t) =>
+  effects.getCurrentDir()
```
Comment (bryphe) [inline src/editor/Model/QuickOpen.re]: It looks like right now we only pull in the top-level directory - one enhancement we'll need down the road is to also filter through subdirectories. Delegating that to a utility like `ripgrep` might be helpful.

File: src/editor/bin/Oni2.re [hunk 4]
```diff
@@ -172,29 +173,40 @@ let init = app => {
...
-  Reglfw.Glfw.glfwSetCharModsCallback(w.glfwWindow, (_w, codepoint, mods) =>
-    switch (Input.charToCommand(codepoint, mods)) {
-    | None => ()
-    | Some(v) =>
+  let keyEventListener = key =>
+    switch (key, Focus.focused) {
+    | (None, _) => ()
+    | (Some("<ESC>" as v), {contents: Some(_)}) =>
+      /**
+       If a Revery UI element is focused but <ESC> is hit this should
```
Comment (bryphe) [inline src/editor/bin/Oni2.re]: 👍

File: src/editor/Core/Filter.re [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/**
+   Filter
+
+   Module to handle filtering of items using various strategies
+ */
+
+let menu = (query, items) =>
+  Types.UiMenu.(
```
Comment (bryphe) [inline src/editor/Core/Filter.re]: Is this a place where we could use @CrossR 's reasonFuzz library? https://github.com/CrossR/reasonFuzz

Comment (Akin909) [reply src/editor/Core/Filter.re]: @bryphe I separated that with exactly that in mind but would rather leave hooking it up to @CrossR this is just place holder rudimentary filtering logic

File: assets/configuration/keybindings.json (hunk 1 above)
Comment (Akin909) [reply assets/configuration/keybindings.json]: yeah they are that's a hang over from me just trying to get it working initially, I'll tweak that

File: src/editor/Model/Menu.re (hunk 2 above)
Comment (Akin909) [reply src/editor/Model/Menu.re]: it was previously called `menuType` it represents the menu's state/type I guess it always has to be something or rather its always theoretically something i.e. `Closed/NoMenu`, `QuickOpen` or `CommandPalette` we change it whenever we open a menu. It tells us its type and we record it in state. The idea being that we can render different UI per menu or use it to specify behaviour per menu type.

It's our only way of knowing what the difference is between the quick open being open or the command palettte

**Comment by Akin909** [comment]
@bryphe switching the event listeners is proving a little trickier than I'd hoped so I'm going to merge this as is and continue working on swapping those out separately

**Pull request merged**